// @cubejs-backend/native/dist/js/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.shutdownInterface = exports.registerInterface = exports.setLogLevel = exports.isSupported = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
function loadNative() {
    // Development version
    if (fs_1.default.existsSync(path_1.default.join(__dirname, '/../../index.node'))) {
        return require(path_1.default.join(__dirname, '/../../index.node'));
    }
    if (fs_1.default.existsSync(path_1.default.join(__dirname, '/../../native/index.node'))) {
        return require(path_1.default.join(__dirname, '/../../native/index.node'));
    }
    throw new Error(`Unable to load @cubejs-backend/native, probably your system (${process.arch}-${process.platform}) with Node.js ${process.version} is not supported.`);
}
function isSupported() {
    return fs_1.default.existsSync(path_1.default.join(__dirname, '/../../index.node')) || fs_1.default.existsSync(path_1.default.join(__dirname, '/../../native/index.node'));
}
exports.isSupported = isSupported;
function wrapNativeFunctionWithChannelCallback(fn) {
    const native = loadNative();
    return async (extra, channel) => {
        try {
            const result = await fn(JSON.parse(extra));
            if (process.env.CUBEJS_NATIVE_INTERNAL_DEBUG) {
                console.debug("[js] channel.resolve", {
                    result
                });
            }
            channel.resolve(JSON.stringify(result));
        }
        catch (e) {
            channel.reject(e.message || 'Unknown JS exception');
            // throw e;
            console.debug("[js] channel.reject", {
                e
            });
        }
    };
}
;
const setLogLevel = (level) => {
    const native = loadNative();
    native.setLogLevel(level);
};
exports.setLogLevel = setLogLevel;
const registerInterface = async (options) => {
    if (typeof options !== 'object' && options == null) {
        throw new Error('Argument options must be an object');
    }
    if (typeof options.checkAuth != 'function') {
        throw new Error('options.checkAuth must be a function');
    }
    if (typeof options.load != 'function') {
        throw new Error('options.load must be a function');
    }
    if (typeof options.meta != 'function') {
        throw new Error('options.meta must be a function');
    }
    const native = loadNative();
    return native.registerInterface({
        ...options,
        checkAuth: wrapNativeFunctionWithChannelCallback(options.checkAuth),
        load: wrapNativeFunctionWithChannelCallback(options.load),
        meta: wrapNativeFunctionWithChannelCallback(options.meta),
    });
};
exports.registerInterface = registerInterface;
const shutdownInterface = async (instance) => {
    const native = loadNative();
    await native.shutdownInterface(instance);
    await new Promise((resolve) => setTimeout(resolve, 2000));
};
exports.shutdownInterface = shutdownInterface;
